addons/binding/org.openhab.binding.solaredge/src/main/java/org/openhab/binding/solaredge/internal/model/AbstractAggregateDataResponsePrivateApi.java

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.solaredge.internal.model;
+
+import java.math.RoundingMode;
+import java.text.DecimalFormat;
+
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * this abstract class is used as base for the specific aggregate response classes
+ *
+ * @author Alexander Friese - initial contribution
+ */
+public abstract class AbstractAggregateDataResponsePrivateApi implements DataResponse {
+
+    private static final String UNIT_WH = "Wh";
+    private static final String UNIT_KWH = "KWh";
+    private static final String UNIT_MWH = "MWh";
+
+    public static class Value {
+        public Double value;
+        public String unit;
+    }
+
+    public static class ValueAndPercent extends Value {
+        public Double percentage;
+    }
+
+    public static class UtilizationMeasures {
+        public Value production;
+        public Value consumption;
+        public ValueAndPercent selfConsumptionForConsumption;
+        public ValueAndPercent batterySelfConsumption;
+        @SerializedName("import")
+        public ValueAndPercent imported;
+        public ValueAndPercent export;
+    }
+
+    private UtilizationMeasures utilizationMeasures;
+
+    protected final String getValueAsKWh(Value value) {
+        Double convertedValue = value.value;
+
+        if (value.unit != null && value.unit.equals(UNIT_WH)) {
```

Comment:
Here you could actually directly benefit from UoM by parsing the value including its unit - there would be no need to convert it to kWh here.

Reply:
Unfortunately UoM is case sensitive and it does not detect KWh as a valid unit. In addition kW is not available so I need to convert to W manually.
But nevertheless I think UoM is a good idea so I will implement this.

Reply:
Maybe you can preprocess the unit and correct "KWh" to "kWh" yourself before passing it to UoM. What do you mean with "kW" not available? That should imho definitely be identified by UoM. @htreu, can you maybe jump in here and help?

Reply:
I will do the unit preprocessing as you suggested.
I found the Metric prefixes which I can use in combination with unit Watt.

Reply:
FTR, I did two small test cases proving the units and parsing succeeds:

```java
@Test
public void testEnergy() {
    QuantityType<Energy> kwh = new QuantityType<>("2 kWh");
    assertThat(kwh, is(new QuantityType<>(2000, SmartHomeUnits.WATT_HOUR)));
    assertThat(new QuantityType<>("2000 Wh"), is(kwh));
}

@Test
public void testEnergy2() {
    QuantityType<Energy> kw = new QuantityType<>("2 kW");
    assertThat(kw, is(new QuantityType<>(2, MetricPrefix.KILO(SmartHomeUnits.WATT))));
}
```

Reply:
UoM implementation is done. I also did some more refactoring to simplify the code (at least this was my goal...). Before thee was a Map<String, String> used to transfer all the values this approach was quite generic but in the handler those values had to be converted again. @Hilbrand had several remarks there. Now I use Map<Channel, State> which means the handler can use the State objects and pass them directly into the updateState method.

Reply:
I have one more question regarding UoM:
If my solarplant delivers for 500W in the early morning and that value is provided as 500 with unit W and later the production is higher and then I get 2,2 with unit kW. What is then stored in the persistence? I guess only the values are stored but if the unit changes frequently it does not make any sense to rely on the persistence for calculations, etc.

Reply:
Persistence services are currently not yet UoM aware (see https://github.com/eclipse/smarthome/pull/5419#issuecomment-400608692). The recommended way is therefore to define a unit in the label in your items file (and not use `%unit%` in there, because this will make sure that the value is converted to this unit and thus the time-series is not corrupted by having values with changing units.

Reply:
Ok, thanks. This is ok for me I just do not want to have corrupted time series in my persistence.